File: service/account.go
package service

import (
	"claude-code-relay/common"
	"claude-code-relay/model"
	"errors"
	"log"
	"time"
)

type AccountService struct{}

func NewAccountService() *AccountService {
	return &AccountService{}
}

// GetAccountList 获取账号列表
func (s *AccountService) GetAccountList(page, limit int, userID *uint) (*model.AccountListResponse, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 10
	}

	accounts, total, err := model.GetAccountList(page, limit, userID)
	if err != nil {
		return nil, errors.New("获取账号列表失败")
	}

	result := &model.AccountListResponse{
		Accounts: accounts,
		Total:    total,
		Page:     page,
		Limit:    limit,
	}

	return result, nil
}

// CreateAccount 创建账号
func (s *AccountService) CreateAccount(req *model.CreateAccountRequest, userID uint) (*model.Account, error) {
	account := &model.Account{
		Name:         req.Name,
		PlatformType: req.PlatformType,
		RequestURL:   req.RequestURL,
		SecretKey:    req.SecretKey,
		GroupID:      req.GroupID,
		Priority:     req.Priority,
		Weight:       req.Weight,
		EnableProxy:  req.EnableProxy,
		ProxyURI:     req.ProxyURI,
		ActiveStatus: req.ActiveStatus,
		IsMax:        req.IsMax,
		AccessToken:  req.AccessToken,
		RefreshToken: req.RefreshToken,
		ExpiresAt:    req.ExpiresAt,
		UserID:       userID,
	}

	if err := model.CreateAccount(account); err != nil {
		return nil, errors.New("创建账号失败")
	}

	return account, nil
}

// GetAccountByID 根据ID获取账号详情
func (s *AccountService) GetAccountByID(id uint, userID *uint) (*model.Account, error) {
	account, err := model.GetAccountByID(id)
	if err != nil {
		return nil, errors.New("账号不存在")
	}

	// 如果指定了用户ID，验证账号是否属于该用户
	if userID != nil && account.UserID != *userID {
		return nil, errors.New("无权访问此账号")
	}

	return account, nil
}

// UpdateAccount 更新账号
func (s *AccountService) UpdateAccount(id uint, req *model.UpdateAccountRequest, userID *uint) (*model.Account, error) {
	account, err := s.GetAccountByID(id, userID)
	if err != nil {
		return nil, err
	}

	// 更新字段
	account.Name = req.Name
	account.PlatformType = req.PlatformType
	account.RequestURL = req.RequestURL
	account.GroupID = req.GroupID
	account.Priority = req.Priority
	account.Weight = req.Weight
	account.EnableProxy = req.EnableProxy
	account.ProxyURI = req.ProxyURI
	account.ActiveStatus = req.ActiveStatus
	account.IsMax = req.IsMax

	if req.SecretKey != "" {
		account.SecretKey = req.SecretKey
	}

	if req.AccessToken != "" {
		account.AccessToken = req.AccessToken
	}

	if req.RefreshToken != "" {
		account.RefreshToken = req.RefreshToken
	}

	if err := model.UpdateAccount(account); err != nil {
		return nil, errors.New("更新账号失败")
	}

	return account, nil
}

// DeleteAccount 删除账号
func (s *AccountService) DeleteAccount(id uint, userID *uint) error {
	account, err := s.GetAccountByID(id, userID)
	if err != nil {
		return err
	}

	if err := model.DeleteAccount(account.ID); err != nil {
		return errors.New("删除账号失败")
	}

	return nil
}

// GetAccountsByUserID 获取指定用户的所有账号
func (s *AccountService) GetAccountsByUserID(userID uint) ([]model.Account, error) {
	accounts, err := model.GetAccountsByUserID(userID)
	if err != nil {
		return nil, errors.New("获取用户账号列表失败")
	}

	return accounts, nil
}

// UpdateAccountActiveStatus 更新账号激活状态
func (s *AccountService) UpdateAccountActiveStatus(id uint, activeStatus int, userID *uint) error {
	account, err := s.GetAccountByID(id, userID)
	if err != nil {
		return err
	}

	account.ActiveStatus = activeStatus

	if err := model.UpdateAccount(account); err != nil {
		return errors.New("更新账号激活状态失败")
	}

	return nil
}

// UpdateAccountCurrentStatus 更新账号当前状态(值不能为3)
func (s *AccountService) UpdateAccountCurrentStatus(id uint, currentStatus int, userID *uint) error {
	if currentStatus == 3 {
		return errors.New("当前状态不能设置为3")
	}

	account, err := s.GetAccountByID(id, userID)
	if err != nil {
		return err
	}

	account.CurrentStatus = currentStatus

	if err := model.UpdateAccount(account); err != nil {
		return errors.New("更新账号当前状态失败")
	}

	return nil
}

// UpdateAccountStatus 根据响应状态码更新账号状态
func (s *AccountService) UpdateAccountStatus(account *model.Account, statusCode int, usage *common.TokenUsage) {
	// 根据状态码设置CurrentStatus
	switch {
	case statusCode == 429:
		// 限流状态
		account.CurrentStatus = 3
	case statusCode > 400:
		// 接口异常
		account.CurrentStatus = 2
	case statusCode == 200 || statusCode == 201:
		// 正常状态
		account.CurrentStatus = 1

		// 请求成功时更新最后使用时间和今日使用次数
		now := time.Now()

		// 判断最后使用时间是否为当天
		if account.LastUsedTime != nil {
			lastUsedDate := time.Time(*account.LastUsedTime).Format("2006-01-02")
			todayDate := now.Format("2006-01-02")

			if lastUsedDate == todayDate {
				// 同一天，使用次数+1
				account.TodayUsageCount++
			} else {
				// 不同天，重置为1
				account.TodayUsageCount = 1
			}
		} else {
			// 首次使用，设置为1
			account.TodayUsageCount = 1
		}

		// 更新token使用量和费用（如果有的话）
		if usage != nil {
			// 计算本次请求的费用
			costResult := common.CalculateCost(usage)
			currentCost := costResult.Costs.Total

			if account.LastUsedTime != nil {
				lastUsedDate := time.Time(*account.LastUsedTime).Format("2006-01-02")
				todayDate := now.Format("2006-01-02")

				if lastUsedDate == todayDate {
					// 同一天，累加各类tokens和费用
					account.TodayInputTokens += usage.InputTokens
					account.TodayOutputTokens += usage.OutputTokens
					account.TodayCacheReadInputTokens += usage.CacheReadInputTokens
					account.TodayCacheCreationInputTokens += usage.CacheCreationInputTokens
					account.TodayTotalCost += currentCost
				} else {
					// 不同天，重置各类tokens和费用
					account.TodayInputTokens = usage.InputTokens
					account.TodayOutputTokens = usage.OutputTokens
					account.TodayCacheReadInputTokens = usage.CacheReadInputTokens
					account.TodayCacheCreationInputTokens = usage.CacheCreationInputTokens
					account.TodayTotalCost = currentCost
				}
			} else {
				// 首次使用，设置各类tokens和费用
				account.TodayInputTokens = usage.InputTokens
				account.TodayOutputTokens = usage.OutputTokens
				account.TodayCacheReadInputTokens = usage.CacheReadInputTokens
				account.TodayCacheCreationInputTokens = usage.CacheCreationInputTokens
				account.TodayTotalCost = currentCost
			}
		}

		// 更新最后使用时间
		nowTime := model.Time(now)
		account.LastUsedTime = &nowTime
	default:
		// 其他状态码保持原状态
		return
	}

	// 更新数据库
	if err := model.UpdateAccount(account); err != nil {
		log.Printf("failed to update account status: %v", err)
	}
}

File: web/src/api/account.ts
import { request } from '@/utils/request';

// API 账号接口
const Api = {
  GetList: '/api/v1/accounts/list',
  Create: '/api/v1/accounts/create',
  Update: '/api/v1/accounts/update',
  Delete: '/api/v1/accounts/delete',
  GetDetail: '/api/v1/accounts/detail',
  UpdateActiveStatus: '/api/v1/accounts/update-active-status',
  UpdateCurrentStatus: '/api/v1/accounts/update-current-status',
  TestAccount: '/api/v1/accounts/test',
  // Claude OAuth 相关
  GetOAuthURL: '/api/v1/oauth/generate-auth-url',
  ExchangeCode: '/api/v1/oauth/exchange-code',
};

// 账号信息
export interface Account {
  id: number;
  name: string;
  platform_type: string; // claude/claude_console/openai/gemini
  request_url: string;
  secret_key: string; // 现在会返回密钥
  access_token: string; // 现在会返回访问令牌
  refresh_token: string; // 现在会返回刷新令牌
  expires_at: number;
  is_max: boolean;
  group_id: number;
  priority: number;
  weight: number;
  today_usage_count: number;
  today_input_tokens: number;
  today_output_tokens: number;
  today_cache_read_input_tokens: number;
  today_cache_creation_input_tokens: number;
  today_total_cost: number;
  enable_proxy: boolean;
  proxy_uri: string;
  model_mapping: string;
  last_used_time: string;
  rate_limit_end_time: string;
  current_status: number; // 1:正常,2:接口异常,3:账号异常/限流
  active_status: number; // 1:激活,2:禁用
  user_id: number;
  created_at: string;
  updated_at: string;
  user: {
    id: number;
    username: string;
  };
  group: {
    id: number;
    name: string;
  } | null;
  // 最近一周统计数据
  weekly_cost: number; // 最近一周使用费用
  weekly_count: number; // 最近一周使用次数
}

// 创建账号参数
export interface AccountCreateParams {
  name: string;
  platform_type: string;
  request_url?: string;
  secret_key?: string;
  group_id?: number;
  priority?: number;
  weight?: number;
  enable_proxy?: boolean;
  proxy_uri?: string;
  model_mapping?: string;
  active_status?: number;
  is_max?: boolean;
  access_token?: string;
  refresh_token?: string;
  expires_at?: number;
  today_usage_count?: number;
}

// 更新账号参数
export interface AccountUpdateParams {
  name: string;
  platform_type: string;
  request_url?: string;
  secret_key?: string;
  group_id?: number;
  priority?: number;
  weight?: number;
  enable_proxy?: boolean;
  proxy_uri?: string;
  model_mapping?: string;
  active_status?: number;
  is_max?: boolean;
  access_token?: string;
  refresh_token?: string;
  today_usage_count?: number;
}

// 账号列表参数
export interface AccountListParams {
  page?: number;
  limit?: number;
  user_id?: number;
}

// 账号列表响应
export interface AccountListResponse {
  accounts: Account[];
  total: number;
  page: number;
  limit: number;
}

// 更新激活状态参数
export interface UpdateAccountActiveStatusParams {
  active_status: number;
}

// 更新当前状态参数
export interface UpdateAccountCurrentStatusParams {
  current_status: number;
}

// 测试账号响应
export interface TestAccountResponse {
  success: boolean;
  message: string;
  status_code: number;
  platform_type: string;
}

// Claude OAuth 相关接口类型

// OAuth授权URL响应
export interface OAuthURLResponse {
  auth_url: string;
  state: string;
  code_challenge: string;
  code_verifier: string;
}

// OAuth授权码验证参数
export interface ExchangeCodeParams {
  authorization_code: string;
  callback_url: string;
  proxy_uri: string;
  code_verifier: string;
  state: string;
}

// OAuth授权码验证响应
export interface ExchangeCodeResponse {
  access_token: string;
  refresh_token: string;
  expires_at: number;
  user_info: {
    id: string;
    email: string;
    name: string;
  };
}

/**
 * 获取账号列表
 */
export function getAccountList(params?: AccountListParams) {
  return request.get<AccountListResponse>({
    url: Api.GetList,
    params,
  });
}

/**
 * 获取账号详情
 */
export function getAccountDetail(id: number) {
  return request.get<Account>({
    url: `${Api.GetDetail}/${id}`,
  });
}

/**
 * 创建账号
 */
export function createAccount(data: AccountCreateParams) {
  return request.post<Account>({
    url: Api.Create,
    data,
  });
}

/**
 * 更新账号
 */
export function updateAccount(id: number, data: AccountUpdateParams) {
  return request.put<Account>({
    url: `${Api.Update}/${id}`,
    data,
  });
}

/**
 * 删除账号
 */
export function deleteAccount(id: number) {
  return request.delete({
    url: `${Api.Delete}/${id}`,
  });
}

/**
 * 批量删除账号
 */
export function batchDeleteAccounts(ids: number[]) {
  return request.delete({
    url: Api.Delete,
    data: { ids },
  });
}

/**
 * 更新账号激活状态
 */
export function updateAccountActiveStatus(id: number, active_status: number) {
  return request.put({
    url: `${Api.UpdateActiveStatus}/${id}`,
    data: { active_status },
  });
}

/**
 * 批量更新账号激活状态
 */
export function batchUpdateAccountActiveStatus(ids: number[], active_status: number) {
  return request.put({
    url: Api.UpdateActiveStatus,
    data: { ids, active_status },
  });
}

/**
 * 更新账号当前状态
 */
export function updateAccountCurrentStatus(id: number, current_status: number) {
  return request.put({
    url: `${Api.UpdateCurrentStatus}/${id}`,
    data: { current_status },
  });
}

/**
 * 批量更新账号当前状态
 */
export function batchUpdateAccountCurrentStatus(ids: number[], current_status: number) {
  return request.put({
    url: Api.UpdateCurrentStatus,
    data: { ids, current_status },
  });
}

// 测试账号
export function testAccount(id: number) {
  return request.post<TestAccountResponse>({
    url: `${Api.TestAccount}/${id}`,
  });
}

// === Claude OAuth 相关接口 ===

/**
 * 获取Claude OAuth授权URL
 */
export function getOAuthURL() {
  return request.get<OAuthURLResponse>({
    url: Api.GetOAuthURL,
  });
}

/**
 * 验证OAuth授权码并获取token
 */
export function exchangeCode(data: ExchangeCodeParams) {
  return request.post<ExchangeCodeResponse>({
    url: Api.ExchangeCode,
    data,
  });
}
